src/common/buffer.cc

```diff
@@ -404,185 +463,381 @@ static simple_spinlock_t buffer_debug_lock = SIMPLE_SPINLOCK_INITIALIZER;
   };
 #endif
 
+
 #ifdef CEPH_HAVE_SPLICE
-  class buffer::raw_pipe : public buffer::raw {
-  public:
-    MEMPOOL_CLASS_HELPERS();
+/*
+ * Specialization of buffer::raw that implements zero-copy techniques.
+ * It is created for linux kernel. Uses syscalls and behaviours specific to it.
+ *
+ * Data is stored in series of linux pipes.
+ * If not specifically required by invoking \ref get_data(), data will remain in kernel space.
+ *
+ * \note For non-linux builds this class is not compiled and buffer::create_zero_copy produces regular buffer.
+ */
+class buffer::raw_pipe : public buffer::raw {
+private:
+  static constexpr int read_end = 0;
+  static constexpr int write_end = 1;
+  static constexpr int min_pipe_size = 1024*256; // < drop from using zero-copy if pipes are too small
+  struct pipe_elem
+  {
+    size_t size;
+    int fd;
+  };
+  std::vector<pipe_elem> pipe_read_elements;
+  int pipe_curr_write_fd; //this changes as writing progresses to next pipes
+  size_t total_content_size;
+  unsigned expected_len;
+
+public:
+  MEMPOOL_CLASS_HELPERS();
+  explicit raw_pipe(unsigned expected_len) : raw(0),
+      pipe_curr_write_fd(-1),
+      total_content_size(0),
+      expected_len(expected_len) {
+    update_max_pipe_size();
+    if (get_max_pipe_size()<min_pipe_size) {
+      data = new char[expected_len];
+      len = expected_len;
+      return;
+    }
+    int no_pipes = expected_len / get_max_pipe_size() + 1; //this is estimated #pipes to be needed
```

Comment:
also could you add a space after `//` in the comment line. so it's easier to read?